### PR TITLE
feat: terraform module set up for C8000V Autonomous mode - device type

### PR DESCRIPTION
**Pull request by kpdhulipala**
feat: terraform module set up for C8000V Autonomous mode - device type

### CODE REVIEW

File: examples/c8000v-ha/README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Network Edge C8000V Router Single Device Example
```
Comment (rling-equinix) [inline examples/c8000v-ha/README.md:1]: This is for HA.

File: examples/c8000v-ha/README.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+| Name                                                          | Version |
+|---------------------------------------------------------------|---------|
+| <a name="provider_equinix"></a> [equinix](#provider\_equinix) | >= 1.34 |
+
+## Modules
+
+| Name                                                   | Source                  | Version |
+|--------------------------------------------------------|-------------------------|---------|
+| <a name="module_c8000v"></a> [c8000v](#module\_c8000v) | ../../../modules/c8000v | n/a     |
```
Comment (rling-equinix) [inline examples/c8000v-ha/README.md:43]: We need to update the version here.

File: examples/c8000v-ha/README.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+## Resources
+
+| Name                                                                                                                                          | Type     |
+|-----------------------------------------------------------------------------------------------------------------------------------------------|----------|
+| [equinix_network_acl_template.c8000v_pri](https://registry.terraform.io/providers/equinix/equinix/latest/docs/resources/network_acl_template) | resource |
```
Comment (rling-equinix) [inline examples/c8000v-ha/README.md:49]: The link here is not correct.

File: examples/c8000v-ha/main.tf [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+module "c8000v_ha" {
+  source               = "../../modules/c8000v"
+  name                 = "tf-c8000v-router-ha"
+  metro_code           = var.metro_code_primary
+  platform             = "small"
+  byol                 = true
+  account_number       = "664566"
+  license_token        = ""
```
Comment (rling-equinix) [inline examples/c8000v-ha/main.tf:13]: If license token is optional, we should ignore this field in the example or set up a default value if this is not passed by customer.

File: modules/c8000v/README.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+|------|------|
+| [equinix_network_device.cluster](https://registry.terraform.io/providers/equinix/equinix/latest/docs/resources/network_device) | resource |
+| [equinix_network_device.non_cluster](https://registry.terraform.io/providers/equinix/equinix/latest/docs/resources/network_device) | resource |
+| [equinix_network_device_platform.this](https://registry.terraform.io/providers/equinix/equinix/latest/docs/data-sources/network_device_platform) | data source |
+| [equinix_network_device_software.this](https://registry.terraform.io/providers/equinix/equinix/latest/docs/data-sources/network_device_software) | data source |
+| [equinix_network_device_type.this](https://registry.terraform.io/providers/equinix/equinix/latest/docs/data-sources/network_device_type) | data source |
+
+## Inputs
+
+| Name | Description | Type | Default | Required |
```
Comment (rling-equinix) [inline modules/c8000v/README.md:51]: License token should also be listed here if it is optional.

File: modules/c8000v/main.tf [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  core_count           = data.equinix_network_device_platform.this.core_count
+  metro_code           = var.metro_code
+  connectivity         = var.connectivity
+  account_number       = var.account_number
+  license_token        = var.byol ? var.license_token : null
+  term_length          = var.term_length
+  interface_count      = var.interface_count
+  notifications        = var.notifications
+  acl_template_id      = var.acl_template_id != "" ? var.acl_template_id : null
+  additional_bandwidth = var.additional_bandwidth > 0 ? var.additional_bandwidth : null
```
Comment (rling-equinix) [inline modules/c8000v/main.tf:43]: The default value should be 0, not null.

File: modules/c8000v/main.tf [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+data "equinix_network_device_software" "this" {
+  device_type = data.equinix_network_device_type.this.code
+  packages    = [var.software_package]
+  stable      = true
+  most_recent = true
+}
+
+resource "equinix_network_device" "non_cluster" {
+  lifecycle {
+    ignore_changes = [version, core_count]
```
Comment (rling-equinix) [inline modules/c8000v/main.tf:20]: Any ideas why those two fields are ignored?

**Review by rling-equinix — CHANGES_REQUESTED** [review]

File: examples/c8000v-ha/README.md (hunk 2 above)
Comment (kpdhulipala) [reply examples/c8000v-ha/README.md:43]: done

File: examples/c8000v-ha/README.md (hunk 3 above)
Comment (kpdhulipala) [reply examples/c8000v-ha/README.md:49]: updated

File: modules/c8000v/README.md (hunk 5 above)
Comment (kpdhulipala) [reply modules/c8000v/README.md:51]: listed

File: modules/c8000v/main.tf (hunk 6 above)
Comment (kpdhulipala) [reply modules/c8000v/main.tf:43]: updated

File: modules/c8000v/main.tf (hunk 7 above)
Comment (kpdhulipala) [reply modules/c8000v/main.tf:20]: removed ignore changes to allow vm resizing.

File: examples/c8000v-ha/README.md (hunk 1 above)
Comment (kpdhulipala) [reply examples/c8000v-ha/README.md:1]: done

File: modules/c8000v/main.tf (hunk 6 above)
Comment (kpdhulipala) [reply modules/c8000v/main.tf:36]: as discussed its going to be as is

File: modules/c8000v/variables.tf [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    error_message = "One of following platform flavors are supported: small, medium, large."
+  }
+}
+
+variable "software_package" {
+  description = "Device software package"
+  type        = string
+  validation {
+    condition     = can(regex("^(network-advantage|network-essentials|network-premier)$", var.software_package))
+    error_message = "One of following software packages are supported: STD."
```
Comment (rling-equinix) [inline modules/c8000v/variables.tf:35]: Could you add more details here for the supported software packages?

File: modules/c8000v/variables.tf [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+  }
+}
+
+variable "name" {
+  description = "Device name"
+  type        = string
+  validation {
+    condition     = length(var.name) >= 2 && length(var.name) <= 50
+    error_message = "Device name should consist of 2 to 50 characters."
```
Comment (rling-equinix) [inline modules/c8000v/variables.tf:44]: I think the default maximum value is 30 characters.

File: modules/c8000v/variables.tf [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+  default     = "INTERNET-ACCESS"
+}
+
+variable "additional_bandwidth" {
+  description = "Additional internet bandwidth for a device"
+  type        = number
+  default     = 0
+  validation {
+    condition     = var.additional_bandwidth == 0 || (var.additional_bandwidth >= 25 && var.additional_bandwidth <= 2001)
+    error_message = "Additional internet bandwidth should be between 25 and 2001 Mbps."
```
Comment (rling-equinix) [inline modules/c8000v/variables.tf:105]: The upper bound should be 5001.

File: modules/c8000v/variables.tf [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+  validation {
+    condition     = !try(var.secondary.enabled, false) || can(regex("^[A-Z]{2}$", var.secondary.metro_code))
+    error_message = "Key 'metro_code' has to be defined for secondary device. Valid metro code consists of two capital leters, i.e. SV, DC."
+  }
+  validation {
+    condition     = !try(var.secondary.enabled, false) || try(length(var.secondary.hostname) >= 2 && length(var.secondary.hostname) <= 10, false)
+    error_message = "Key 'hostname' has to be defined for secondary device. Valid hostname has to be from 2 to 10 characters long."
+  }
+  validation {
+    condition     = !try(var.secondary.enabled, false) || try(var.secondary.additional_bandwidth >= 25 && var.secondary.additional_bandwidth <= 2001, true)
```
Comment (rling-equinix) [inline modules/c8000v/variables.tf:139]: Upper bound needs to be updated.

File: modules/c8000v/variables.tf [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+  validation {
+    condition     = length(var.name) >= 2 && length(var.name) <= 50
+    error_message = "Device name should consist of 2 to 50 characters."
+  }
+}
+
+variable "hostname" {
+  description = "Device hostname"
+  type        = string
+  default     = ""
```
Comment (rling-equinix) [inline modules/c8000v/variables.tf:51]: Can we add the validation here? I think hostname is a required field and has length limit.

File: modules/c8000v/variables.tf [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+variable "license_token" {
+  description = "License token"
+  type        = string
+  default     = ""
```
Comment (rling-equinix) [inline modules/c8000v/variables.tf:57]: Shall we remove the default value for optional field?

**Review by rling-equinix — CHANGES_REQUESTED** [review]

File: examples/c8000v-ha/README.md (hunk 1 above)
Comment (kpdhulipala) [reply examples/c8000v-ha/README.md]: done

File: modules/c8000v/variables.tf (hunk 9 above)
Comment (kpdhulipala) [reply modules/c8000v/variables.tf:44]: I tested , its min 3 to 50 chars

File: modules/c8000v/variables.tf (hunk 12 above)
Comment (kpdhulipala) [reply modules/c8000v/variables.tf]: done

File: modules/c8000v/variables.tf (hunk 13 above)
Comment (kpdhulipala) [reply modules/c8000v/variables.tf:60]: I tried using optional identifier but its valid for only object type, only way is to go with setting null as default

File: modules/c8000v/variables.tf [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+variable "project_id" {
+  description = "project_id"
+  type        = string
+}
+
+variable "account_number" {
+  description = "Billing account number for a device"
+  type        = string
+  default     = 0
```
Comment (rling-equinix) [inline modules/c8000v/variables.tf:18]: Can we add a similar validation rule as acl template for account number?

**Review by rling-equinix — CHANGES_REQUESTED** [review]

File: modules/c8000v/variables.tf (hunk 14 above)
Comment (kpdhulipala) [reply modules/c8000v/variables.tf]: done

**Review by rling-equinix — APPROVED** [review]

**Review by schadalawada — APPROVED** [review]

**Pull request merged**